src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java

```diff
@@ -39,30 +47,35 @@ public SSLEngineResult wrap(final ByteBuffer[] srcs, final int offset, final int
 
     @Override
     public SSLEngineResult unwrap(final ByteBuffer src, final ByteBuffer[] dsts, final int offset, final int length) throws SSLException {
-        final SSLEngineResult result;
-        if (generatedJa3Signature) {
-            result = engine.unwrap(src, dsts, offset, length);
-        } else {
-            // Ensure no-one modifies the buffer underneath as per sun.security.ssl.SSLEngineImpl
-            String ja3 = null;
-            final HandshakeStatus handshakeStatus = engine.getHandshakeStatus();
-            if (HandshakeStatus.NOT_HANDSHAKING != handshakeStatus && HandshakeStatus.FINISHED != handshakeStatus) {
-                ja3 = new JA3Signature().ja3Signature(src);
-            } else {
-                generatedJa3Signature = true;
-            }
-
-            result = engine.unwrap(src, dsts, offset, length);
-
-            if (ja3 != null) {
-                // no lock needed, it is fine if we write same digest multiple times
-                engine.getSession().putValue(JA3Constants.JA3_FINGERPRINT, ja3);
-                generatedJa3Signature = true;
-            }
-        }
-
-
-        return result;
+		final SSLEngineResult result;
```

Comment:
Would this work?

    @Override
    public SSLEngineResult unwrap(final ByteBuffer src, final ByteBuffer[] dsts, final int offset, final int length) throws SSLException {
		if (!ja3Done) {
		    if (ja3Signature != null) {
		        final SSLSession handshakeSession = engine.getHandshakeSession();
		        if (handshakeSession != null) {
		            // Set ja3 signature in handshake session
		            handshakeSession.putValue(JA3Constants.JA3_FINGERPRINT, ja3);
		            ja3Done = true;
		        }
		    } else {
		        // 1. Generate JA3 signature
		        final HandshakeStatus handshakeStatus = engine.getHandshakeStatus();
		        if (HandshakeStatus.FINISHED == handshakeStatus) {
				    ja3Done = true;
			    } else {
			        ja3 = new JA3Signature().ja3Signature(src);
			    }
		    }
		}

		return engine.unwrap(src, dsts, offset, length);
    }